File: counting-valleys/counting-valleys.go
package countingvalleys

func countingValleys(_ int32, directions string) int32 {
	var position int32 = 0
	var valleyCount int32 = 0
	inValley := false

	for _, dir := range directions {
		if dir == 'U' {
			position++
		} else {
			position--
		}

		if position < 0 {
			inValley = true
		}

		if position == 0 && inValley {
			inValley = false
			valleyCount++
		}
	}

	return valleyCount
}

File: counting-valleys/CountingValleys.java
public class CountingValleys {

  public static int countingValleys(int ignored, String directions) {
    int position = 0;
    int valleyCount = 0;
    boolean inValley = false;

    for (int i = 0; i < directions.length(); i++) {
      char direction = directions.charAt(i);
      if (direction == 'U') {
        position++;
      } else {
        position--;
      }

      if (position < 0) {
        inValley = true;
      }

      if (position == 0 && inValley) {
        inValley = false;
        valleyCount++;
      }
    }

    return valleyCount;
  }

}
